File: head_to_offer/32-1/hto_32_1.py
from typing import List, Dict
from queue import Queue
from collections import deque
import math
import itertools

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return []
        def bfs(node: TreeNode, level: int)-> None:
            if level >= len(res):
                res.append([])
            res[level].append(node.val)
            if node.left is not None:
                bfs(node.left, level+1)
            if node.right is not None:
                bfs(node.right, level+1)
        res = []
        bfs(root, 0)
        return res

File: head_to_offer/32-1/hto_32_1.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Queue;
import java.util.ArrayDeque;
import java.util.LinkedList;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}


class Solution {
	public List<List<Integer>> levelOrder(TreeNode root) {
		List<List<Integer>> res = new ArrayList<>();
		if (root == null) return res;
		Queue<TreeNode> queue = new LinkedList<>();
	//	Queue<TreeNode> queue = new ArrayDeque<>();
		queue.add(root);
		while (!queue.isEmpty()) {
			int count = queue.size();
			List<Integer> resIn = new ArrayList<>();
			while (count-- > 0) {
				TreeNode t = queue.poll();
				resIn.add(t.val);
				if (t.left  != null) queue.add(t.left);
				if (t.right != null) queue.add(t.right);
			}
			res.add(resIn);
		}
		return res;
	}
}


public class leetcode {
	public static void main(String[] args) {
		TreeNode a = new TreeNode(3);
		TreeNode b = new TreeNode(9);
		TreeNode c = new TreeNode(20);
		TreeNode d = new TreeNode(15);
		TreeNode e = new TreeNode(7);
		a.left = b;
		a.right = c;
		c.left = d;
		c.right = e;
		Solution s = new Solution();
		List<List<Integer>> res = s.levelOrder(a);
		for (List<Integer> l : res) {
			for (Integer ll : l)
				System.out.print(l + " ");
		}
	}
}
